hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java

```diff
@@ -778,6 +820,74 @@ private String planIfLast(AbstractSpliceFunction f, boolean isLast) {
         return (plan != null && !plan.isEmpty() ? plan : f.getSparkName());
     }
 
+    public static boolean joinProjectsOnlyLeftTableColumns(JoinType joinType) {
+        return (joinType == DataSet.JoinType.LEFTANTI ||
+                joinType == DataSet.JoinType.LEFTSEMI);
+    }
+
+    private static boolean
+    leftDFOfJoinProjectsFewerColumnsThanDefined(JoinOperation op,
+                                                JoinType      joinType,
+                                                Dataset<Row>  leftDF) {
+        return !joinProjectsOnlyLeftTableColumns(joinType) &&
+               !op.wasRightOuterJoin &&
+                op.getLeftRow().nColumns() > leftDF.schema().length();
+    }
+
+    // If the left DataSet defines more columns than the left DataFrame,
+    // the column names will be off.  Let's fill the gaps to match the
+    // ExecRowDefinition.
+    private Dataset<Row> fixupColumnNames(JoinOperation op, JoinType joinType,
+                                          Dataset<Row> leftDF, Dataset<Row> rightDF,
+                                          Dataset<Row> joinedDF) {
+        boolean needsFixup =
+                leftDFOfJoinProjectsFewerColumnsThanDefined(op, joinType, leftDF);
```

Comment:
Why would we have a situation where leftDF project fewer columns than defined?

Reply:
I talked to @msirek , one scenario would be a regular join whose left source is a semi join. For derby, if it is a semi join, the immediate join row includes columns from the right even if it is not needed(these columns are pruned later), for spark, the columns from the right table of the semi-join won't be projected to the join result row. This difference could cause leftDF of the regular join to project few columns (as it is semi-join result computed from spark) than defined (which is determined in Derby).